boot/cypress/cy_flash_pal/cy_flash_psoc6.c

```diff
@@ -0,0 +1,277 @@
+* above is prohibited without the express written permission of Cypress.
+*
+* Disclaimer: THIS SOFTWARE IS PROVIDED AS-IS, WITH NO
+* WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING,
+* BUT NOT LIMITED TO, NONINFRINGEMENT, IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+* PARTICULAR PURPOSE. Cypress reserves the right to make
+* changes to the Software without notice. Cypress does not assume any
+* liability arising out of the application or use of the Software or any
+* product or circuit described in the Software. Cypress does not
+* authorize its products for use in any products where a malfunction or
+* failure of the Cypress product may reasonably be expected to result in
+* significant property damage, injury or death ("High Risk Product"). By
+* including Cypress's product in a High Risk Product, the manufacturer
+* of such system or application assumes all risk of such use and in doing
+* so agrees to indemnify Cypress against all liability.
+*
+******************************************************************************/
+#include "string.h"
+#include "stdlib.h"
+#include "stdbool.h"
+
+#ifdef MCUBOOT_HAVE_ASSERT_H
+#include "mcuboot_config/mcuboot_assert.h"
+#else
+#include <assert.h>
+#endif
+
+#include "cy_device_headers.h"
+
+// TODO:
+//#include "device.h"
+//#include "util.h"
+//#include "flash.h"
+//#include "flash_map.h"
+#include "cy_flash_psoc6.h"
+
+#include "cy_flash.h"
+#include "cy_syspm.h"
+
+#define PSOC6_WR_SUCCESS 0
+#define PSOC6_WR_ERROR_INVALID_PARAMETER 1
+#define PSOC6_WR_ERROR_FLASH_WRITE 2
+
+#define PSOC6_FLASH_ERASE_BLOCK_SIZE	CY_FLASH_SIZEOF_ROW /* PSoC6 Flash erases by Row */
+
+int psoc6_flash_read(off_t addr, void *data, size_t len)
+{
+	/* flash read by simple memory copying */
+	memcpy((void *)data, (const void*)addr, (size_t)len);
+
+	return 0;
+}
+
+int psoc6_flash_write(off_t addr,
+			    const void *data, size_t len)
+{
+	int rc;
+
+	rc = psoc6_flash_write_hal((uint8_t *)data, addr, len);
+
+	return rc;
+}
+
+int psoc6_flash_erase(off_t addr, size_t size)
+{
+	int rc = 0;
+
+	uint32_t addrStart, addrEnd, address;
+	uint32_t remStart, remEnd;
+	uint32_t rowIdxStart, rowIdxEnd, rowNum;
+	uint8_t  buff[CY_FLASH_SIZEOF_ROW];
+
+	addrStart = addr;
+	addrEnd   = addrStart + size;
+
+	/* find if area bounds are aligned to rows */
+	remStart = addrStart%CY_FLASH_SIZEOF_ROW;
+	remEnd   = addrEnd%CY_FLASH_SIZEOF_ROW;
+
+	/* find which row numbers are affected for full Erase */
+	rowIdxStart = addrStart/CY_FLASH_SIZEOF_ROW;
+	rowIdxEnd   = addrEnd/CY_FLASH_SIZEOF_ROW;
+
+	if(remStart != 0)
+	{/* first row is fragmented, move to next */
+		rowIdxStart++;
+	}
+
+	/* total number of rows for full erase */
+	rowNum = rowIdxEnd - rowIdxStart;
+	address = rowIdxStart*CY_FLASH_SIZEOF_ROW;
+
+	while(rowNum>0)
+	{
+		rc = Cy_Flash_EraseRow(address);
+		assert(rc == 0);
+		address += CY_FLASH_SIZEOF_ROW;
+		rowNum--;
+	}
+
+	/* if Start of erase area is unaligned */
+	if(remStart != 0)
```

Comment:
This really should just be an error. MCUboot should never be erasing regions that are misaligned, and by reading them, erasing, and then writing, you create a window where an untimely powerdown will result in a loss of data.

If you are finding the code calling with misaligned erases, we need to figure out what is happening and fix that.